Make sprite picking opt-in

chompaa opened the pull request:
# Objective

Fixes #16903.

## Solution

- Make sprite picking opt-in by requiring a new `SpritePickingCamera` component for cameras and usage of a new `Pickable` component for entities.
- Update the `sprite_picking` example to reflect these changes.
- Some reflection cleanup (I hope that's ok).

## Testing

Ran the `sprite_picking` example

## Open Questions

<del>
   <ul>
    <li>Is the name `SpritePickable` appropriate?</li>
    <li>Should `SpritePickable` be in `bevy_sprite::prelude?</li>
  </ul> 
</del>

## Migration Guide

The sprite picking backend is now strictly opt-in using the `SpritePickingCamera` and `Pickable` components. You should add the `Pickable` component any entities that you want sprite picking to be enabled for, and mark their respective cameras with `SpritePickingCamera`.

alice-i-cecile commented on crates/bevy_sprite/src/picking_backend.rs:
```diff
@@ -39,17 +46,27 @@ pub struct SpritePickingSettings {
+/// Only needed if [`SpritePickingSettings::require_markers`] is set to `true`, and ignored otherwise.
+#[derive(Debug, Clone, Default, Component, Reflect)]
+#[reflect(Debug, Default, Component)]
+pub struct SpritePickable;
```
Part of me feels like this should be backend agnostic as `Pickable` 🤔

chompaa replied:
I'm inclined to agree, but what if there existed a scenario where multiple backends are in use, and you want to opt-in to just one of them for a specific entity?

alice-i-cecile replied:
Each entity should only be covered by one backend. I don't think we should try to plan for extremely pathological cases like that.

alice-i-cecile commented on crates/bevy_sprite/src/picking_backend.rs:
```diff
@@ -60,14 +77,21 @@ impl Plugin for SpritePickingPlugin {
     sprite_query: Query<(
         Entity,
         &Sprite,
+        Option<&SpritePickable>,
```
Don't we need to use a `With` filter to actually get the desired performance benefits?

alice-i-cecile commented on crates/bevy_sprite/src/picking_backend.rs:
```diff
@@ -30,6 +31,12 @@ pub enum SpritePickingMode {
+    ///
+    /// This setting is provided to give you fine-grained control over which cameras and entities
+    /// should be used by the sprite picking backend at runtime.
+    pub require_markers: bool,
```
I don't like this field: it adds a surprising behavior change with little benefit. We should pick a standard option, and then stick to it.

chompaa commented on crates/bevy_sprite/src/picking_backend.rs:
```diff
@@ -60,14 +77,21 @@ impl Plugin for SpritePickingPlugin {
+    cameras: Query<(
+        Entity,
+        &Camera,
+        Option<&SpritePickable>,
```
Agreed, I think there should be another marker component. Not sure on a name, though. `SpritePickingCamera` comes to mind.

alice-i-cecile reviewed (changes requested):
Looking better :) Final changes:

1. Move marker component to bevy_picking, and call it `Pickable`.
2. Split apart the components and call the camera one `SpritePickingCamera`.

Follow-up:

- use this design for mesh-picking as well

chompaa commented:
>     2. Split apart the components and call the camera one `SpritePickingCamera`.

What are your thoughts on a `PickingCamera` component in bevy_picking? I can see it being used in the mesh picking backend and the UI picking backend if we were to support diegetic UI.

I'd be happy to tackle the follow-up :)

chompaa commented:
> What are your thoughts on a `PickingCamera` component in bevy_picking? I can see it being used in the mesh picking backend and the UI picking backend if we were to support diegetic UI.

On second thought, I don't think this makes sense. Cameras could be used for distinct backends, so generalizing seems wrong.

alice-i-cecile commented:
> On second thought, I don't think this makes sense. Cameras could be used for distinct backends, so generalizing seems wrong.

Exactly my thinking.

aevyrie commented on crates/bevy_picking/src/backend.rs:
```diff
@@ -41,11 +41,19 @@ use bevy_reflect::Reflect;
+/// You should use this component if you are using a backend with opt-in behavior.
+#[derive(Component, Reflect, Debug)]
+#[reflect(Component, Debug)]
+pub struct Pickable;
```
This type existed in mod_picking before upstream. It then turned into `PickingBehavior`, an optional component. We should discuss before adding this back.

aevyrie commented on crates/bevy_picking/src/backend.rs:
```diff
@@ -41,11 +41,19 @@ use bevy_reflect::Reflect;
+/// You should use this component if you are using a backend with opt-in behavior.
+#[derive(Component, Reflect, Debug)]
+#[reflect(Component, Debug)]
+pub struct Pickable;
```
This type existed in mod_picking before upstream. It then turned into PickingBehavior, an optional component. We should discuss before adding this back.

aevyrie replied:
I'd rather backends use the existing component. E.g. if a backed is set to opt-in, it should be looking for the `PickingBehavior` component.

https://docs.rs/bevy_mod_picking/latest/bevy_mod_picking/backend/prelude/struct.Pickable.html

That said, I'm not a huge name of the rename to `PickingBehavior`. I much prefer `Pickable` as a component that makes an entity pickable.

alice-i-cecile replied:
I agree that we should unify this. No strong feelings on the naming.

chompaa replied:
My initial thoughts were that these were describing two distinct things, but I see that converging on one makes sense now. I have a preference towards `Pickable` since I think it better conveys that it will be considered in the backend.

alice-i-cecile commented on crates/bevy_sprite/src/picking_backend.rs:
```diff
@@ -1,6 +1,11 @@
+//! # Usage
+//!
+//! This backend is strictly opt-in. For entities to be considered for sprite picking, you should
+//! mark them with [`Pickable`] and their respective cameras with [`SpritePickingCamera`].
```
Hmm, the debugging use case is a good argument for making this configurable 🤔 I would like to unify this across all of our backends though.

chompaa replied:
My only qualm is that we lose some performance benefit, but I'm in agreement.

chompaa replied:
There is the question of whether it should be opt in by default or not. Backend consistency says it should be opt out, but perhaps we want to push users to opt in?

aevyrie replied:
From my experience working on mod_picking, it's a better experience for users if it starts as opt-out, because it remove friction and steps to just get things working. Once it is working, it's then another small step to switch to opt-in if they decide to.

chompaa commented:
I've pushed some changes:

- Removed `Pickable` in favor of using `PickingBehavior` (I'd like to follow up on a rename).
- Added `require_markers` to `SpritePickingSettings` (defaults to `false` to align with the mesh picking backend).

I also think that the `SpritePickingPlugin` should be explicitly added by the user, and not a part of `DefaultPlugins` (and in general, all backends should be), but that seems out of scope for this PR.

aevyrie commented:
> I've pushed some changes:
> 
> * Removed `Pickable` in favor of using `PickingBehavior` (I'd like to follow up on a rename).
> * Added `require_markers` to `SpritePickingSettings` (defaults to `false` to align with the mesh picking backend).
> 
> I also think that the `SpritePickingPlugin` should be explicitly added by the user, and not a part of `DefaultPlugins` (and in general, all backends should be), but that seems out of scope for this PR.

Agree with everything here. Doing a pass to standardize all backends and make them all opt-in would be great.
